frontend/src/main/web/src/components/TextInput.jsx

```diff
@@ -0,0 +1,185 @@
+import React, { Component, PropTypes } from 'react'
+import ReactDOM from 'react-dom'
+import TextareaAutosize from 'react-textarea-autosize'
+import { flattenThemeClasses } from '../utils/styleUtils'
+import { isEqual } from 'lodash'
+
+const classes = {
+  base: {
+    ap: 'Ap(n)',
+    bgc: 'Bgc(t)',
+    bdc: 'Bdc(neutral)',
+    bdw: 'Bdw(2px)',
+    bds: 'Bds(s)',
+    bdrs: 'Bdrs(rq)',
+    bxz: 'Bxz(bb)',
+    c: 'C(i)',
+    ff: 'Ff(inh)',
+    fw: 'Fw(inh)',
+    fz: 'Fz(inh)',
+    fs: 'Fs(inh)',
+    o: 'O(n)',
+    p: 'Px(rq) Py(re)',
+    ph: 'Ph(neutral)',
+    focus: {
+      bdc: 'Bdc(pri):f'
+    }
+  }
+}
+
+class TextInput extends Component {
+  constructor () {
+    super()
+  }
+
+  _onBlur (e) {
+    const { onBlur } = this.props
+    if (onBlur) {
+      onBlur(e)
+    }
+  }
+  _onChange (e) {
+    const { onChange, onChangeText } = this.props
+    if (onChangeText) onChangeText(e.target.value)
+    if (onChange) onChange(e)
+  }
+
+  _onFocus (e) {
+    const { clearTextOnFocus, onFocus, selectTextOnFocus } = this.props
+    const node = ReactDOM.findDOMNode(this)
+    if (clearTextOnFocus) node.value = ''
+    if (selectTextOnFocus) node.select()
+    if (onFocus) onFocus(e)
+  }
+
+  _onSelectionChange (e) {
+    const { onSelectionChange } = this.props
+    const { selectionDirection, selectionEnd, selectionStart } = e.target
+    const event = {
+      selectionDirection,
+      selectionEnd,
+      selectionStart,
+      nativeEvent: e.nativeEvent
+    }
+    if (onSelectionChange) onSelectionChange(event)
+  }
+
+  _onKeyDown (e) {
+    const { onKeyDown } = this.props
+    if (onKeyDown) onKeyDown(e)
+  }
+
+  _onClear () {
+    const node = ReactDOM.findDOMNode(this)
+    node.value = ''
+  }
+
+  render () {
+    const {
+      id,
+      accessibilityLabel,
+      autoComplete,
+      autoFocus,
+      defaultValue,
+      editable = true,
+      keyboardType = 'default',
+      maxLength,
+      maxNumberOfLines,
+      multiline = false,
+      numberOfLines = 2,
+      onBlur,
+      onChange,
+      onChangeText,
+      onKeyDown,
+      onSelectionChange,
+      placeholder,
+      secureTextEntry = false,
+      theme,
+      value
+    } = this.props
+
+    let type
+
+    switch (keyboardType) {
+      case 'email-address':
+        type = 'email'
+        break
+      case 'numeric':
+        type = 'number'
+        break
+      case 'phone-pad':
+        type = 'tel'
+        break
+      case 'url':
+        type = 'url'
+        break
+    }
```

Comment:
Looking at the other code, this should always hit one of the cases. Add a `default` case that will log an error since if it does that it means the code is broken (and we want to know about it ASAP).